agent/eventstream/eventstream.go

```diff
@@ -0,0 +1,129 @@
+// permissions and limitations under the License.
+
+// Package handler deals with appropriately reacting to all ACS messages as well
+// as maintaining the connection to ACS.
+package eventstream
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/cihub/seelog"
+	"golang.org/x/net/context"
+)
+
+type eventHandler func(...interface{}) error
+
+// EventStream waiting for events and notifying the listeners by invoking
+// the handler that listeners registered
+type EventStream struct {
+	name         string
+	open         bool
+	event        chan interface{}
+	handlers     map[string]eventHandler
+	ctx          context.Context
+	handlersLock sync.RWMutex
+	statusLock   sync.RWMutex
+}
+
+func NewEventStream(name string, ctx context.Context) *EventStream {
+	return &EventStream{
+		name:     name,
+		event:    make(chan interface{}, 1),
+		ctx:      ctx,
+		open:     false,
+		handlers: make(map[string]eventHandler),
+	}
+}
+
+// Subscribe adds the handler to be called into EventStream
+func (eventStream *EventStream) Subscribe(name string, handler eventHandler) error {
+	eventStream.handlersLock.Lock()
+	defer eventStream.handlersLock.Unlock()
+
+	if _, ok := eventStream.handlers[name]; ok {
+		return fmt.Errorf("handler name has already existed, %v", name)
+	}
+
+	eventStream.handlers[name] = handler
+	return nil
+}
+
+// broadcast calls all handler's handler function
+func (eventStream *EventStream) broadcast(event interface{}) {
+	eventStream.handlersLock.RLock()
+	defer eventStream.handlersLock.RUnlock()
+
+	seelog.Debugf("Event stream %v received events, broadcasting to listeners...", eventStream.name)
+
+	for _, handlerFunc := range eventStream.handlers {
+		go handlerFunc(event)
+	}
+}
+
+// Unsubscribe deletes the handler from the EventStream
+func (eventStream *EventStream) Unsubscribe(name string) {
+	eventStream.handlersLock.Lock()
+	defer eventStream.handlersLock.Unlock()
+
+	for handler, _ := range eventStream.handlers {
+		if handler == name {
+			seelog.Debugf("Unsbuscribe event handler %v from event stream %v", handler, eventStream.name)
+			delete(eventStream.handlers, handler)
+			return
+		}
+	}
+}
+
+// WriteToEventStream writes event to the event stream
+func (eventStream *EventStream) WriteToEventStream(event interface{}) error {
+	eventStream.statusLock.RLock()
+	defer eventStream.statusLock.RUnlock()
+
+	if !eventStream.open {
+		return fmt.Errorf("Event stream is closed")
+	}
+	eventStream.event <- event
+	return nil
+}
+
+// Context returns the context of event stream
+func (eventStream *EventStream) Context() context.Context {
+	return eventStream.ctx
+}
+
+// StartListening start listening to the event stream and handle received event
+func (eventStream *EventStream) StartListening() {
+	seelog.Infof("Event stream %v start listening...", eventStream.name)
+	eventStream.statusLock.Lock()
+	defer eventStream.statusLock.Unlock()
+	eventStream.open = true
+
+	go func() {
```

Comment:
Can you move this to a named function? Those are easier to read in a stack trace (and you're not closing over anything defined in this outer function, so you don't need it to be anonymous or pass anything in).